### DISCUSSION TITLE
HMAC with external provider

### DISCUSSION THREAD

**Discussion by vchandap** [openssl/openssl#25484] [Q&A] [answered]
Hi,

I am trying to calculate hmac with external OpenSSL provider (HSM), having secret key in HSM, 
how to pass key id with openssl dgst command

**Comment by paulidale** [comment]
`-propquery provider="your provider"` is the way to go here.  You might also need `-provider <your provider>` to load it properly.

**Reply by vchandap** [reply]
But how pass 'keyid' as openssl dgst command doesn't have option like -inkey?

**Reply by vchandap** [reply]
Hi,

I was asking another way: does the OpenSSL digest provider allow an already provisioned key through provider-keymgmt?

**Reply by t8m** [reply]
No, OpenSSL APIs do not yet have an abstraction for symmetric keys. So it is currently not possible. In theory a HMAC provider for HSM could somehow workaround it - for example by using a NULL key and an additional parameter specifying the key id but that would not be a general solution.

**Comment by t8m** [comment, marked as answer]
There is an RFE here https://github.com/openssl/openssl/issues/19754

**Reply by vchandap** [reply]
Thank you for the info.
